File: javascript/packages/linter/src/rules/a11y-no-aria-unsupported-elements.ts
import { ParserRule } from "../types.js"
import { ElementStackVisitor } from "./rule-utils.js"

import { getAttributeName } from "@herb-tools/core"

import type { ParseResult, ParserOptions, HTMLAttributeNode } from "@herb-tools/core"
import type { UnboundLintOffense, LintContext, FullRuleConfig } from "../types.js"

class NoAriaUnsupportedElementsVisitor extends ElementStackVisitor {
  private UNSUPPORTED_ELEMENTS = new Set([
    "html",
    "meta",
    "script",
    "style",
  ])

  visitHTMLAttributeNode(node: HTMLAttributeNode): void {
    if (!this.currentTagName || !this.UNSUPPORTED_ELEMENTS.has(this.currentTagName)) return

    const attributeName = getAttributeName(node)
    if (!attributeName) return
    if (!attributeName.startsWith("aria-") && attributeName !== "role") return

    this.addOffense(
      `The \`${attributeName}\` attribute is not supported on the \`<${this.currentTagName}>\` element. ARIA roles, states, and properties should not be used on elements that are not visible or not interactive.`,
      node.location,
    )

    super.visitHTMLAttributeNode(node)
  }
}

export class A11yNoAriaUnsupportedElementsRule extends ParserRule {
  static ruleName = "a11y-no-aria-unsupported-elements"
  static introducedIn = this.version("unreleased")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: false,
      severity: "warning"
    }
  }

  get parserOptions(): Partial<ParserOptions> {
    return {
      action_view_helpers: true,
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new NoAriaUnsupportedElementsVisitor(this.ruleName, context)

    visitor.visit(result.value)

    return visitor.offenses
  }
}

File: rust/herb-linter/src/rules/a11y_no_aria_unsupported_elements.rs
use crate::offense::UnboundOffense;
use crate::rule::{LintContext, ParserRule, Rule};
use crate::utils::element_stack::ElementStack;
use crate::utils::tag_utils::{get_attribute_name, get_tag_local_name};

use herb::nodes::{HTMLAttributeNode, HTMLElementNode};
use herb::ParseResult;
use herb::Visitor;
use herb_config::{Severity, SeverityConfig};

const UNSUPPORTED_ELEMENTS: &[&str] = &["html", "meta", "script", "style"];

pub struct A11yNoAriaUnsupportedElementsRule;

struct NoAriaUnsupportedElementsVisitor {
  rule_name: &'static str,
  offenses: Vec<UnboundOffense>,
  element_stack: ElementStack,
}

impl Visitor for NoAriaUnsupportedElementsVisitor {
  fn visit_html_element_node(&mut self, node: &HTMLElementNode) {
    self.element_stack.push_optional(get_tag_local_name(node));
    self.walk_html_element_node(node);
    self.element_stack.pop();
  }

  fn visit_html_attribute_node(&mut self, node: &HTMLAttributeNode) {
    let current_tag_name = match self.element_stack.current_tag_name() {
      Some(tag_name) if UNSUPPORTED_ELEMENTS.contains(&tag_name) => tag_name.to_string(),
      _ => return,
    };

    let attribute_name = match get_attribute_name(node) {
      Some(name) => name,
      None => return,
    };

    if !attribute_name.starts_with("aria-") && attribute_name != "role" {
      return;
    }

    self.offenses.push(UnboundOffense::new(
      self.rule_name,
      format!(
        "The `{}` attribute is not supported on the `<{}>` element. ARIA roles, states, and properties should not be used on elements that are not visible or not interactive.",
        attribute_name, current_tag_name
      ),
      node.location.clone(),
    ));

    self.walk_html_attribute_node(node);
  }
}

impl Rule for A11yNoAriaUnsupportedElementsRule {
  fn name(&self) -> &'static str {
    "a11y-no-aria-unsupported-elements"
  }

  fn default_severity(&self) -> SeverityConfig {
    SeverityConfig::Severity(Severity::Warning)
  }

  fn default_enabled(&self) -> bool {
    false
  }

  fn parser_options(&self) -> herb::ParserOptions {
    herb::ParserOptions {
      action_view_helpers: true,
      ..crate::rule::default_linter_parser_options()
    }
  }
}

impl ParserRule for A11yNoAriaUnsupportedElementsRule {
  fn check(&self, result: &ParseResult, _context: &LintContext) -> Vec<UnboundOffense> {
    let mut visitor = NoAriaUnsupportedElementsVisitor {
      rule_name: self.name(),
      offenses: Vec::new(),
      element_stack: ElementStack::new(),
    };

    visitor.visit_document_node(&result.value);

    visitor.offenses
  }
}
